engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    /**
+     * Maps the provided function over any expression defined as a child of the
+     * node this is called on.
+     *
+     * @param fn the function to transform the expressions
+     * @return `this`, potentially having had its children transformed by `fn`
+     */
+    IR mapExpressions(Function<Expression, Expression> fn);
+
+    /**
+     * Gets the list of all children IR nodes of this node.
+     *
+     * @return this node's children.
+     */
+    List<IR> children();
+
+    /**
+     * Lists all the nodes in the preorder walk of the tree of this node.
+     *
+     * @return all the descendants of this node.
+     */
+    default List<IR> preorder() {
+        List<IR> ordered = children().flatMap(c -> c.preorder());
+        IR element = this;
+        return $colon$colon$.MODULE$.apply(element, ordered); // ordered.prepended(element) is reporeted as ambiguous
+    }
+
+    /**
+     * Pretty prints the IR.
+     *
+     * @return a pretty-printed representation of the IR
+     */
+    default String pretty() {
+        return Debug.pretty(this.toString());
+    }
+
+    /**
+     * Gets the node's identifier.
+     *
+     * @return the node's identifier
+     */
+    default IR.Identifier getId() {
+        return id();
+    }
+
+    /**
+     * A unique identifier for a piece of IR.
+     */
+    IR.Identifier id();
+
+    /**
+     * Storage for compiler diagnostics related to the IR node.
+     */
+    DiagnosticStorage diagnostics();
+
+    /**
+     * Creates a deep structural copy of `this`, representing the same structure.
+     * <p>
+     * You can choose to keep the location, metadata and diagnostic information
+     * in the duplicated copy, as well as whether you want to generate new
+     * node identifiers or not.
+     *
+     * @param keepLocations   whether locations should be kept in the duplicated IR
+     * @param keepMetadata    whether the pass metadata should be kept in the duplicated IR
+     * @param keepDiagnostics whether the diagnostics should be kept in the duplicated IR
+     * @param keepIdentifiers whether the identifiers should be regenerated in the duplicated IR
+     * @return a deep structural copy of `this`
+     */
+    IR duplicate(
+            boolean keepLocations,
+            boolean keepMetadata,
+            boolean keepDiagnostics,
+            boolean keepIdentifiers
+    );
+
+    /**
+     * Shows the IR as code.
+     *
+     * @param indent the current indentation level
+     * @return a string representation of `this`
+     */
+    String showCode(int indent);
+
+    default String showCode() {
+        return showCode(0);
+    }
+
+
+    /** Creates a random identifier.
+     *
+     * @return a random identifier
+     */
+    static IR.Identifier randomId() {
+        return new Identifier(UUID.randomUUID());
+    }
+
+    /** The type of identifiers for IR nodes. */
+    record Identifier(UUID id) implements Serializable {}
```

Comment:
This way we are certainly incompatible with previous serialized format. Moreover we always serialize one object more. Shouldn't we work directly with `UUID` rather than having `Identifier` and `ExternalId` wrapper. If we want to differentiate between these two classes, then Java offers _type annotations_ - e.g. we could have `@Identifier UUID id` and `@ExternalID UUID id` and with a bit of help of `javac` those wouldn't mix. Here is [more info](https://docs.oracle.com/javase/tutorial/java/annotations/type_annotations.html) - I admit I never deeply looked at the [checker framework](https://checkerframework.org/).

Reply:
`Identifier` and `ExternalId` were previously type aliases which erased to `UUID`. I know that we will allocate an additional record in every case but couldn't think of anything better.

_type annotations_ seem interesting, haven't thought of that. I do wonder how will that play with the rest of Scala though. I will check it.

Reply:
After doing some exploration I don't think this will work in the mixed Scala/Java world. Annotations in Scala don't play well with Java ones.